### ISSUE TITLE
Objects within Schema

### ISSUE THREAD

**Issue by MickL** [szokodiakos/typegoose#54]
I want to do the following (regular Mongoose):

```
let personSchema = new mongoose.Schema({
  id: String,
  address: {
    street: String,
    streetnumber: Number,
  }
});
```

With Typegoose:

```
class Person extends Typegoose {
  @prop()
  id: string;

 @prop()
  address: {
    street: string;
    streetnumber: number;
  }
});
```

But it says: `Object is not a primitive type nor a Typegoose schema (Not extending it).`

If i do like described in the readme(creating a seperate class for address) Typegoose would create a new schema and a seperate document for address and that is not what i want to do.

**Comment by szokodiakos** [comment]
Hello, thanks for your interest in Typegoose. Yes I'm afraid currently sub schemas are supported only. This is an already requested feature so it is on the list.

**Comment by 19majkel94** [comment]
@szokodiakos I'm afraid you won't be able to do that. Typescript reflection is very limited and the only way to have sub documents is by defining sub schemas using classes.

**Comment by MickL** [comment]
Any estimation for this? Is this complicated and do you have time for this enhancement?

**Comment by szokodiakos** [comment]
@MickL Sorry I don't have the time for research and implementation right now. Currently I'm just trying to fix critical issues, and review other folks PRs.

@19majkel94 that is bad news, then maybe there isn't a viable solution currently at all. 😞

**Comment by yann-kaiser** [comment]
@MickL there's a workaround that i just found.

You could create a class for your address and use it in person.

**Comment by yann-kaiser** [comment]
@MickL there's a workaround that i just found.

You could create a class for your address and use it in person.
All you have to do, is cast it to an object while setting it before saving.

Here's an example:

```TS
class Address {
  public street: string
  public streetnumber: number
}

class Person extends Typegoose {
  @prop()
  id: string;

 @prop()
  address: Address
});

export let PersonModel = new Person().getModelForClass(Address)
```

```TS
// Here the code before saving

const person = new PersonModel()
const address = new Address()
address.street = "some street"
address.streetnumber = 42
person.address = Object.assign({}, address)
```

No need to cast it back to Address when loading, the only trick is the ```Object.assign({}, address)``` which creates an object containing the fields in address.

Hope that helps you!

**Comment by richarddd** [comment]
This issue should be closed? The work around is to add a new class and use that class as type. Another option is the use `object` (works with latest typegoose)

**Comment by dainbrump** [comment]
What if I have a Ref in a nested object? For example, I have an Environment model:

```javascript
class EnvironmentModel extends Typegoose {
  @prop()
  name: string;
  //... other properties below
}
```

and then I have an Organization model that references that Environment model in a nested object. The desired schema is:
```javascript
const OrganizationSchema = new mongoose.Schema({
  name: { type: String, required: true, index: { unique: true } },
  suborganizations: [{ type: mongoose.Schema.Types.ObjectId, ref: 'Organization' }],
  settings: {
    environments: [{ type: mongoose.Schema.Types.ObjectId, ref: 'Environment' }]
  }
} // exported with mongoose.model('Organization', OrganizationSchema)
```
I've tried the following:
```javascript
class OrgSettings {
  environment: Ref<EnvironmentModel>[]
}

class OrganizationModel extends Typegoose {
  @prop({required: true, index: true, unique: true })
  name: string;

  @arrayProp({ itemsRef: OrganizationModel })
  suborganizations: Ref<OrganizationModel>[];

  @prop()
  settings: OrgSettings
}
```

This doesn't appear to work and I'm curious if I'm misunderstanding something.

**Comment by AlecDusheck** [comment]
Anyone come up with a good solution for the problem mentioned above?

**Comment by hasezoey** [comment]
This seems to work in version 6.0.0

**Comment by hasezoey** [comment]
@Ben305 this can be closed because this should work in 5.9.0+
